torch/csrc/jit/fuser/cuda/fused_kernel.cpp

```diff
@@ -0,0 +1,187 @@
+, const std::vector<PartitionDesc> _concat_desc
+, const bool _has_random)
+: FusedKernel{_name, _code, _input_desc, _output_desc, _chunk_desc, _concat_desc, _has_random} 
+, device_{_device} {
+  // Initializes driver's API context (if necessary)
+  CUcontext pctx = 0;
+  TORCH_CU_CHECK(cuCtxGetCurrent(&pctx));
+  if (!pctx) {
+     std::unique_lock<std::mutex> cudaFreeMutexLock(
+     *(THCCachingAllocator_getCudaFreeMutex()));
+     cudaFree(0);
+  }
+
+  // Note: hacked at::DeviceGuard since at::DeviceGuard was failing to work
+  // properly in some scenarios
+  int prior_device;
+  cudaGetDevice(&prior_device);
+  cudaSetDevice(device_);
+  
+  // Acquires device and NVRTC properties (for compile arch and occupancy calculations)
+  TORCH_CUDA_CHECK(cudaGetDeviceProperties(&prop_, device_)); 
+  int major, minor;
+  getMajorMinor(prop_, major, minor);
+
+  // Creates the NVRTC program
+  nvrtcProgram program;
+  TORCH_NVRTC_CHECK(nvrtcCreateProgram(
+    &program
+  , code_.c_str()
+  , nullptr
+  , 0
+  , nullptr
+  , nullptr));
+  
+  const std::string compute = "--gpu-architecture=compute_" + std::to_string(major) + std::to_string(minor);
+  const std::vector<const char *> args = {"--std=c++11", compute.c_str(), "-default-device"};
+  const auto result = nvrtcCompileProgram(program, args.size(), args.data());
+  if (result == NVRTC_ERROR_COMPILATION) {
+    size_t logsize;
+    nvrtcGetProgramLogSize(program, &logsize);
+    std::vector<char> log(logsize);
+    nvrtcGetProgramLog(program, log.data());
+    std::stringstream cu;
+    cu << log.data();
+    throw std::runtime_error(cu.str());
+  }
+  ResourceGuard holdProgram([&] {
+    TORCH_NVRTC_CHECK(nvrtcDestroyProgram(&program));
+  });
+  TORCH_NVRTC_CHECK(result);
+  size_t ptx_size;
+  TORCH_NVRTC_CHECK(nvrtcGetPTXSize(program, &ptx_size));
+  ptx_.resize(ptx_size);
+  TORCH_NVRTC_CHECK(nvrtcGetPTX(program, ptx_.data()));
+
+  TORCH_CU_CHECK(cuModuleLoadData(&module_, ptx_.data()));
+  TORCH_CU_CHECK(cuModuleGetFunction(&function_, module_, name_.c_str()));
+
+  // Computes max blocks
+  TORCH_CU_CHECK(cuOccupancyMaxActiveBlocksPerMultiprocessor(
+    &maxBlocks_, function_, 128, 0));
+  maxBlocks_ *= prop_.multiProcessorCount;
+
+  // Resets device (end of hacked at::DeviceGuard)
+  cudaSetDevice(prior_device);
+}
+
+static int ceilDiv(const int a, const int b) {
+  return (a + b - 1) / b;
+}
+
+void FusedKernelCUDA::launch_raw(
+  const uint32_t numel
+, std::vector<void*>& arguments) const {
+  at::DeviceGuard{device_};
+  // Hacked at::DeviceGuard (see note above)
+  int prior_device;
+  cudaGetDevice(&prior_device);
+  cudaSetDevice(device_);
+
+  const auto nBlocks = std::min(maxBlocks_, ceilDiv(numel, kBlockSize));
+
+  // Adds random state to arguments if necessary
+  // Note: offset defined here so its lifetime extends to the launch
+  uint64_t offset;
+  if (has_random_) {
+    const auto rand_offset = 4 * (std::ceil(numel / (4.0 * kBlockSize * nBlocks)) + 1);
```

Comment:
Is the +1 at the end necessary given std::ceil? (It's not your code probably)

Reply:
You are correct this is historic and I hadn't considered its correctness. May not be necessary.